netplugin/nameserver/nameserver.go

```diff
@@ -0,0 +1,705 @@
+				Class: dns.ClassINET, Ttl: 64}
+			return r
+		}()
+		rr = append(rr, r)
+	}
+
+	return rr, len(rr)
+}
+
+func (dt *dnsTables) lookUpEndPointV4Record(nameList map[string]bool, name string) ([]dns.RR, int) {
+	rr := []dns.RR{}
+	for i := range nameList {
+		if record, ok := dt.endpointTbl[i]; ok {
+			if record.v4Record.IsUnspecified() != true {
+				r := func() dns.RR {
+					r := new(dns.A)
+					r.A = record.v4Record
+					r.Hdr = dns.RR_Header{Name: name + ".", Rrtype: dns.TypeA,
+						Class: dns.ClassINET, Ttl: 64}
+					return r
+				}()
+				rr = append(rr, r)
+				if len(rr) >= 10 {
+					return rr, len(rr)
+				}
+			}
+		}
+	}
+
+	return rr, len(rr)
+}
+
+func (dt *dnsTables) lookUpEndPointV6Record(nameList map[string]bool, name string) ([]dns.RR, int) {
+	rr := []dns.RR{}
+	for i := range nameList {
+		if record, ok := dt.endpointTbl[i]; ok {
+			if record.v6Record.IsUnspecified() != true {
+				r := func() dns.RR {
+					r := new(dns.AAAA)
+					r.AAAA = record.v6Record
+					r.Hdr = dns.RR_Header{Name: name + ".", Rrtype: dns.TypeAAAA,
+						Class: dns.ClassINET, Ttl: 64}
+					return r
+				}()
+
+				rr = append(rr, r)
+				if len(rr) > 10 {
+					return rr, len(rr)
+				}
+			}
+		}
+	}
+
+	return rr, len(rr)
+}
+
+func (ens *NetpluginNameServer) getBucket(key string) *tenantBucket {
+	hasher := fnv.New32a()
+	hasher.Write([]byte(key))
+	sum := hasher.Sum32()
+	return &ens.buckets[uint(sum)%ens.bucketSize]
+}
+
+func (ens *NetpluginNameServer) getTenantName(key string) string {
+	s := strings.Split(key, ".")
+	return s[len(s)-1]
+}
+
+func (ens *NetpluginNameServer) getEpgName(key string) string {
+	s := strings.Split(key, ":")
+	return s[0]
+}
+
+// AddLbService adds a LB service in non-multi tenant record
+func (ens *NetpluginNameServer) AddLbService(tenant string, name string, v4name string) {
+	if len(v4name) > 0 {
+		if tenant == commonK8sTenant {
+			ens.commonSvc.Set(name, nameRecord{v4Record: net.ParseIP(v4name)})
+		} else {
+			mc := mastercfg.CfgServiceLBState{
+				Tenant:      tenant,
+				ServiceName: name,
+				IPAddress:   v4name,
+			}
+			ens.addService(&mc)
+		}
+	}
+}
+
+// DelLbService deletes LB service from non-multi tenant record
+func (ens *NetpluginNameServer) DelLbService(tenant string, name string) {
+	if tenant == commonK8sTenant {
+		ens.commonSvc.Remove(name)
+	} else {
+		mc := mastercfg.CfgServiceLBState{
+			Tenant:      tenant,
+			ServiceName: name,
+		}
+		ens.addService(&mc)
+	}
+
+}
+
+func (ens *NetpluginNameServer) addService(s core.State) {
+	svc, ok := s.(*mastercfg.CfgServiceLBState)
+	if !ok {
+		ens.incTenantErrStats("", "serviceAdd")
+		return
+	}
+
+	dnsLog.Infof("[tenant: %s]add service name:%s, ipadddr: %s",
+		svc.Tenant, svc.ServiceName, svc.IPAddress)
+	tenant := svc.Tenant
+	tenMap := ens.getBucket(tenant)
+	tenMap.RWMutex.Lock()
+	defer tenMap.RWMutex.Unlock()
+	tenantTables, ok := tenMap.tenantTables[tenant]
+	if !ok {
+		tenantTables = new(dnsTables)
```

Comment:
Can we a function to initialize all tenant tables? The initialization seems to be done at multiple places.